### PR TITLE
[9.0] Add oauth_provider module

### PR DESCRIPTION
**Pull request by Garamotte**
A new module that allows to use Odoo as OAuth provider.
The OAuth protocol is used to delegate authentication, and authorization to access some resources.

This module really need security reviews/tests, as I'm not fully comfortable with the OAuh protocol.

The RFC defines four profiles: Implicit grant, Authorization code grant, Password credentials grand and User credentials grant.
Only Implicit grant and Authorization code grant are implemented yet, and I don't plan to implement the others for now. The work should not be very hard, as most parts are common, but I don't actually need these profiles, and don't have time to do it for the moment.

Note: I created the PR to allow you to begin testing the module.
@pedrobaeza said that this repo will be split into several smaller repos, this PR will need to be recreated on one of the new repos when they are created.

### CODE REVIEW

**Comment by Garamotte** [comment]
Travis errors unrelated (from database_cleanup and auth_signup_verify_email modules).

**Comment by Garamotte** [comment]
Last little thing done (and rebased on 9.0).
You can add replace the WIP label by needs review.

File: oauth_provider/controllers/main.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+            'oauth_provider.authorization', {
+                'oauth_client': client.name,
+                'oauth_scopes': oauth_scopes,
+            })
+
+    @http.route(
+        '/oauth2/authorize', type='http', auth='user', methods=['POST'])
+    def authorize_post(self, *args, **kwargs):
+        """
+        Redirect to the requested URI during the authorization
```
Comment (lasley) [inline oauth_provider/controllers/main.py]: Not a blocker or anything, just FYI - Docstrings should start on the same line that the quote is started on. The first line (in a multi) should be a complete summary, followed by a blank line, followed by more description if necessary. Check out [PEP-257](https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings) for more

File: oauth_provider/controllers/main.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+        oauth2_server = token.client_id.get_oauth2_server()
+
+        # Retrieve needed arguments for oauthlib methods
+        uri = http.request.httprequest.base_url
+        http_method = http.request.httprequest.method
+        body = oauthlib.common.urlencode(
+            http.request.httprequest.values.items())
+        headers = http.request.httprequest.headers
+
+        valid, oauthlib_request = oauth2_server.verify_request(
```
Comment (lasley) [inline oauth_provider/controllers/main.py]: This token & argument retrieval followed verification block seems to be repeated a lot. It would be nice if it was just a helper method instead.

File: oauth_provider/oauth2/validator.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+class OdooValidator(RequestValidator):
+    """
+    OAuth2 validator to be used in Odoo
+    """
+    def authenticate_client(self, request, *args, **kwargs):
+        """
+        Authenticate the client
+        """
+        raise NotImplementedError
+        # return False
```
Comment (lasley) [inline oauth_provider/oauth2/validator.py]: Please remove this commented code

File: oauth_provider/oauth2/validator.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+class OdooValidator(RequestValidator):
+    """
+    OAuth2 validator to be used in Odoo
+    """
+    def authenticate_client(self, request, *args, **kwargs):
+        """
+        Authenticate the client
+        """
+        raise NotImplementedError
```
Comment (lasley) [inline oauth_provider/oauth2/validator.py]: No client auth implementation?

File: oauth_provider/oauth2/validator.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+        # return False
+
+    def authenticate_client_id(self, client_id, request, *args, **kwargs):
+        """
+        Ensure client_id belong to a non-confidential client
+        """
+        request.client = http.request.env['oauth.provider.client'].search([
+            ('identifier', '=', client_id),
+        ])
+        request.odoo_user = http.request.env['oauth.provider.client'].env.user
```
Comment (lasley) [inline oauth_provider/oauth2/validator.py]: Why the `env` and `user` from `oauth.provider.client`? It's the same `env` and `user` isn't it, or is there some fancy context switching that I am missing?

File: oauth_provider/oauth2/validator.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+                ('refresh_token', '=', token),
+            ])
+            db_token.refresh_token = False
+
+    def rotate_refresh_token(self, request):
+        """
+        Determine if the refresh token has to be renewed
+        Called after refreshing an access token
+        """
+        return True
```
Comment (lasley) [inline oauth_provider/oauth2/validator.py]: This doesn't seem very useful, or is it an interface for child classes?

File: oauth_provider/oauth2/validator.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+        """
+        Store the authorization code into the database
+        """
+        redirect_uri = http.request.env['oauth.provider.redirect.uri'].search([
+            ('name', '=', request.redirect_uri),
+        ])
+        http.request.env['oauth.provider.authorization.code'].sudo().create({
+            'code': code['code'],
+            'client_id': request.client.id,
+            'user_id': request.odoo_user.id,
```
Comment (lasley) [inline oauth_provider/oauth2/validator.py:168]: Is it ever possible for this to be hit without the `authenticate_client_id` being called? (real question)

File: oauth_provider/oauth2/validator.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+        return request.client.identifier == client_id and set(scopes).issubset(
+            set(request.client.mapped('scope_ids.code')))
+
+    def validate_user(
+            self, username, password, client, request, *args, **kwargs):
+        """
+        Ensure the usernamd and password are valid
+        """
+        raise NotImplementedError
+        # return False
```
Comment (lasley) [inline oauth_provider/oauth2/validator.py]: Remove commented code

File: oauth_provider/security/oauth_provider_security.xml [hunk 9]
```diff
@@ -0,0 +1,53 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<openerp>
+    <data>
+        ##############################################################################
```
Comment (lasley) [inline oauth_provider/security/oauth_provider_security.xml]: Wait... how is this not breaking XML? Please use the shorter header though, and put just under the xml version. This is how I do it https://github.com/OCA/vertical-medical/blob/9.0/medical/views/medical_menu.xml#L3 - although we don't have specific guidelines in OCA on the XML part AFAIK

File: oauth_provider/security/oauth_provider_security.xml [hunk 10]
```diff
@@ -0,0 +1,53 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<openerp>
```
Comment (lasley) [inline oauth_provider/security/oauth_provider_security.xml]: `<odoo>` and can remove data tag

File: oauth_provider/templates/authorization.xml [hunk 11]
```diff
@@ -0,0 +1,60 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
```
Comment (lasley) [inline oauth_provider/templates/authorization.xml]: Odoo tag and remove data

File: oauth_provider/templates/authorization.xml [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    <data>
+        ##############################################################################
```
Comment (lasley) [inline oauth_provider/templates/authorization.xml]: Same note regarding use of smaller header

File: oauth_provider/tests/test_oauth_provider_client.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    def test_grant_response_type_fields(self):
+        """
+        Check the value of the grant_type and response_type fields
+        """
+        # Default : Web Application
+        client = self.new_client({'identifier': 'default'})
+        self.assertEqual(client.grant_type, 'authorization_code')
+        self.assertEqual(client.response_type, 'code')
+
+        # Web Application
```
Comment (lasley) [inline oauth_provider/tests/test_oauth_provider_client.py]: IMO the scope of all these tests is too large. It's best to make smaller test methods in order to provide a better view of things when stuff is broken vs. execution stopping on the first broken test.

**Review by lasley — CHANGES_REQUESTED** [review]
Awesome thanks @sylvain-garancher - comments inline. Also my bad on leaving the WIP tag for so long, missed the comment.

File: oauth_provider/controllers/main.py (hunk 1 above)
Comment (Garamotte) [reply oauth_provider/controllers/main.py]: Didn't know that, thanks.
I'll have to change my comments on many other PRs :D

File: oauth_provider/oauth2/validator.py (hunk 4 above)
Comment (Garamotte) [reply oauth_provider/oauth2/validator.py]: This file is an implementation of the interface provided by the `oauthlib` module.
As the user is already logged on Odoo, we have the user in the `request.env`, so I didn't need to implement this method (unless I miss something).

File: oauth_provider/oauth2/validator.py (hunk 3 above)
Comment (Garamotte) [reply oauth_provider/oauth2/validator.py]: I leaved this line here to remind the return value's type, but I may add this in the docstring.

File: oauth_provider/oauth2/validator.py (hunk 5 above)
Comment (Garamotte) [reply oauth_provider/oauth2/validator.py]: I think this is a bad copy/paste from the previous line, I'll fix that :)

File: oauth_provider/oauth2/validator.py (hunk 6 above)
Comment (Garamotte) [reply oauth_provider/oauth2/validator.py]: This is the implementation of an interface.
The default value in the interface is True, si this is not needed, but I preferred to implement all methods. We can remove this without any issue, if it is better.

File: oauth_provider/controllers/main.py (hunk 2 above)
Comment (Garamotte) [reply oauth_provider/controllers/main.py]: You're right, I'll provide a helper method (which will be useful for modules depending on this one too).

File: oauth_provider/oauth2/validator.py (hunk 7 above)
Comment (Garamotte) [reply oauth_provider/oauth2/validator.py:168]: Yes, because the user is already logged in Odoo (the first route has `auth='user'`).
The `authenticate_client_id` should be required for the two unimplemented profiles (which need to privide credentials on the first request).

File: oauth_provider/security/oauth_provider_security.xml (hunk 10 above)
Comment (Garamotte) [reply oauth_provider/security/oauth_provider_security.xml]: Oops, I kept the old templates for my XML files :)

File: oauth_provider/security/oauth_provider_security.xml (hunk 9 above)
Comment (Garamotte) [reply oauth_provider/security/oauth_provider_security.xml]: This is part of our own templates since at least v5, I think this is time to update our templates.
I'll change that.

File: oauth_provider/oauth2/validator.py (hunk 6 above)
Comment (lasley) [reply oauth_provider/oauth2/validator.py]: Can we instead just add a note about the method being primarily for inheritance by child classes in the docstring? This is typically what Odoo SA does with interfaces I think, and would make it more clear for a module developer doesn't understand all the base code.

File: oauth_provider/oauth2/validator.py (hunk 3 above)
Comment (lasley) [reply oauth_provider/oauth2/validator.py]: Docstring would be the preferred please 😄

File: oauth_provider/oauth2/validator.py (hunk 7 above)
Comment (lasley) [reply oauth_provider/oauth2/validator.py:168]: Sweet I had a feeling I was missing a piece from the auth workflow. Thanks for the explanation

File: oauth_provider/controllers/main.py (hunk 1 above)
Comment (lasley) [reply oauth_provider/controllers/main.py]: Hah it's one of those things that you can't unsee once you know it, right?!

File: oauth_provider/oauth2/validator.py (hunk 4 above)
Comment (lasley) [reply oauth_provider/oauth2/validator.py]: Nope nope, was definitely me missing something. Can we add a docstring note that this is an interface then?

File: oauth_provider/tests/test_oauth_provider_client.py (hunk 13 above)
Comment (Garamotte) [reply oauth_provider/tests/test_oauth_provider_client.py]: You advise to split in one test method per possible value ?

I left all in a single method, because every test in this method are really close, but I don't know all best practices about tests.

Comment (lasley) [reply oauth_provider/tests/test_oauth_provider_client.py]: Yeah the smaller the test the better. I go to great lengths to follow a 1:1 method:assertion ratio, but sometimes it's not possible given context and setup required. 

Consider an example using these tests. If the response_type for the web application is wrong, we won't know if it's right for Mobile, Legacy, and Backend because the test will stop execution at the first failed assertion. 

It's incredibly valuable to have the whole picture when debugging failing tests because it can help point us closer to the actual error, or specific edge case that is being debugged (maybe only X of the Y are failing).

Comment (Garamotte) [reply oauth_provider/tests/test_oauth_provider_client.py]: Ok, I understand. I'll split the tests when it is possible ;)

**Comment by Garamotte** [comment]
@lasley I updated the PR with all of your comments.

I also created a common class for both controllers tests (initialization and many tests are the same for all controllers).
Is it done right ? I prefixed all methods by an `_` because Odoo ran tests from the common class in addition to running them from the specific classes. I'd like to avoid the need to redefine each common test in the custom test classes, if possible, but didn't found how to do that.

**Comment by Garamotte** [comment]
Rebased on 9.0, but runbot seems broken.
Is it caused by the Docker build ? Do I need to update some configuration files in the repo ?

**Review by lasley — APPROVED** [review]
@sylvain-garancher LGTM OCA/runbot-addons#111 for Runbot issue, which is definitely not related to anything you did

**Review by moylop260 — APPROVED** [review]
LGTM 
I'll review the runbot issue

**Comment by Garamotte** [comment]
I'll finally have to implement the password credentials grant (token request by providing username and password), and add it in this PR.

**Comment by Garamotte** [comment]
A short reminder for myself : Check if everything works well with multiple databases (unauthenticated routes may not work if no database has been selected).

**Comment by Garamotte** [comment]
I also have a PoC for using [JWT](https://jwt.io) instead of a randomly generated token.

Can I add it in this module, or provide another module later to add this feature ?
It requires another python dependency (pyjwt), which is not required for simple OAuth2 authentication.

**Comment by lasley** [comment]
IMO JWT should be a separate module. Not everyone is going to want it (although I do and can't wait :P).

**Comment by Garamotte** [comment]
@lasley If you want to take a look, I added the JWT module here: https://github.com/syleam/server-tools/tree/9.0-add-oauth_provider-jwt-module

I'll create for the new module a PR after this one is merged, but I still have some bugs to fix here.

**Comment by Garamotte** [comment]
I re-added WIP in the PR name.
I have some issues with the web application profile, and I'm currently developing the legacy application profile.

**Comment by Garamotte** [comment]
Waiting for tests by the customer, but the Legacy application profile (password credentials grant type) is now implemented.

I'm not very proud of the `test_full_process` methods (in each controllers test class), but everything else should be ok.

**Comment by Garamotte** [comment]
Tests improvements:
- Better split controller's test stuff from real test methods
- Added `login` and `logout` to avoid the need to mock the env in each method
- I finally found how to include test methods in separate classes without the need to call them explicitely in the real test class

I also fixed the web and legacy application behaviours (`/oauth2/token` using POST, etc.)...
Still needs tests to confirm that this now works as expected in OAuth's RFC.

**Comment by Garamotte** [comment]
New changes:
- Specified HTTP status code for errors (400/401)
- Updated tests to check the HTTP status code after each request

**Comment by Garamotte** [comment]
Tests split is finished, I think this is good now, so I removd the WIP tag in title.
Still waiting for a confirmation from my customer for a potential cors issue.

**Comment by Garamotte** [comment]
After a rebase on 9.0, Travis is failing on many other modules, but not on this module's tests.

**Comment by Garamotte** [comment]
After many tests, my customer has no new issue now, using the "legacy application" profile.
The cors issue doesn't seem to be linked to this module.

**Review by lasley — APPROVED** [review]
Thanks @sylvain-garancher - code looks good. Also confirmed that the CI fail isn't you and is instead #689

**Comment by Garamotte** [comment]
I have an issue with the default values for UUID fields (`oauth_identifier` on `res.users`).
The default looks like this : https://github.com/OCA/server-tools/pull/611/files#diff-4a2885ce924a60c01c46e75c1663da33R14

When installing the module, all existing users get the same UUID as identifier... which is not what we want. How could we generate a new random UUID for each user ?
This is because the default_get method is only called once for all existing records : https://github.com/odoo/odoo/blob/5f63878ff6af844b6eab8d233a3f52ee7622eccd/openerp/models.py#L2396-L2419

I initially though about a `pre_init_hook` which could create the column in database, then populate it (to allow the unique constraint be added on the first run), but there will still be an issue if a script creates users in batches.
Maybe overriding the create method in addtion to a `pre_init_hook` ? Any other idea ?

**Comment by lasley** [comment]
We circumvented this in medical by moving the default generation into an overrided create method instead. Looks like that would still work here? Basiccally:


```
class TestClass:
  def create(self, vals):
    vals['unique_key'] = uuid4()
    return super(TestClass, self).create(vals)
```

Downfall of this is from a UI perspective, because the default won't exist until save. We considered it not a problem in our context, and I don't think it would be here either - the val is read-only for humans anyways.

**Comment by Garamotte** [comment]
I added the `pre_init_hook` for now, as our main issue was the initial install of the module, kept in a separate commit for the review, will be squashed later.

I also added the missing `copy=False` attribute on the two uuid fields (`oauth_identifier` on `res.users` and `identifier` on `oauth.provider.client`). Creating user accounts with the `auth_signup` module is done by duplicating the template user, and all new users had the same identifier.

Do you still see a need for overriding the `create` method ?
When creating several users at once, either the identifier will be in data (and if the same value appears multiple times, the unique constraint will raise), or the identifier will not be in data (in this case, the default function should be called for each new user).

**Comment by lasley** [comment]
@> Do you still see a need for overriding the create method ?

Nope, I'd say you covered the bases. We ended up having to implement the create I think due to layers and layers of delegate inheritance at play & some connectors that bulk add data in ways a user normally wouldn't.

Looks like lint is 🔴  though. The other two failures are known, one of which is being resolved in #719 & the other should clear itself per Pedro's email earlier.

**Comment by Garamotte** [comment]
Lint fixed, rebased on 9.0 to benefit from the tests fixes of the database_cleanup module.
Everything looks good.

**Comment by Garamotte** [comment]
Rebased on 9.0 to resolve the requirements.txt conflict.
This module is now used in production by our customer, since a couple of weeks, without any issue :)

**Review by lasley — APPROVED** [review]
We've been in prod for a few weeks now too. Excellent work, @sylvain-garancher

**Comment by Garamotte** [comment]
Rebased on 9.0 to fix the `requirements.txt` conflict.

**Comment by lasley** [comment]
Thanks for another badass module @sylvain-garancher
